src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandlerParallelOperations.cs

```diff
@@ -0,0 +1,466 @@
+                    expectedNumberOfResources: totalNumberOfRequests);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "BundleHandler - There was an error while initializing a new Bundle Operation: {ErrorMessage}", ex.Message);
+                throw;
+            }
+
+            // Parallel Resource Handling Function.
+            Func<ResourceExecutionContext, CancellationToken, Task> handleRequestFunction = async (ResourceExecutionContext resourceExecutionContext, CancellationToken ct) =>
+            {
+                if (resourceExecutionContext.Index > 0 && resourceExecutionContext.Index % GCCollectTrigger == 0)
+                {
+                    RunGarbageCollection();
+                }
+
+                _logger.LogTrace("BundleHandler - Running '{HttpVerb}' Request #{RequestNumber} out of {TotalNumberOfRequests}.", httpVerb, resourceExecutionContext.Index, totalNumberOfRequests);
+
+                // Creating new instances per record in the bundle, and making their access thread-safe.
+                // Avoiding possible internal conflicts due the parallel access from multiple threads.
+                ResourceIdProvider resourceIdProvider = new ResourceIdProvider();
+                IFhirRequestContext originalFhirRequestContext = (IFhirRequestContext)_originalFhirRequestContext.Clone();
+
+                Stopwatch watch = Stopwatch.StartNew();
+
+                EntryComponent entry = await HandleRequestAsync(
+                    responseBundle,
+                    httpVerb,
+                    throttledEntryComponent,
+                    _bundleType,
+                    bundleOperation,
+                    resourceExecutionContext.Context,
+                    resourceExecutionContext.Index,
+                    resourceExecutionContext.PersistedId,
+                    _requestCount,
+                    auditEventTypeMapping,
+                    originalFhirRequestContext,
+                    requestContext,
+                    bundleHttpContextAccessor,
+                    resourceIdProvider,
+                    fhirJsonParser,
+                    _logger,
+                    ct);
+
+                statistics.RegisterNewEntry(httpVerb, resourceExecutionContext.Index, entry.Response.Status, watch.Elapsed);
+
+                await SetResourceProcessingStatusAsync(httpVerb, resourceExecutionContext, bundleOperation, entry, cancellationToken);
+
+                watch.Stop();
+                _logger.LogTrace("BundleHandler - '{HttpVerb}' Request #{RequestNumber} completed with status code '{StatusCode}' in {TotalElapsedMilliseconds}ms.", httpVerb, resourceExecutionContext.Index, entry.Response.Status, watch.ElapsedMilliseconds);
+            };
+
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            List<Task> requestsPerResource = new List<Task>();
+            foreach (ResourceExecutionContext resourceContext in _requests[httpVerb])
```

Comment:
At this point I left it as a regular foreach just for simplicity.

As I'm accumulating multiple functions running in parallel, doing that using a Parallel.ForEachAsync does not have too much advantage (because the function will be handled in parallel anyway). Additionally, as I'm creating multiple tasks in the process at the same time for each bundle, I've decided keeping this part simple and avoid the creation of even more tasks.